core-bundle/src/Twig/Loader/ContaoFilesystemLoader.php

```diff
@@ -0,0 +1,438 @@
+     *
+     * @param string $name The template logical name
+     *
+     * @throws LoaderError When $name is not found
+     */
+    public function getSourceContext($name): Source
+    {
+        $templateName = $this->getThemeTemplateName($name) ?? $name;
+
+        $source = parent::getSourceContext($templateName);
+
+        // The Contao PHP templates will still be rendered by the Contao
+        // framework via a PhpTemplateProxyNode. We're removing the source to
+        // not confuse Twig's lexer and parser and just keep the block names.
+        // At some point we may transpile the source to valid Twig instead and
+        // drop the proxy.
+        if ('html5' !== Path::getExtension($source->getPath(), true)) {
+            return $source;
+        }
+
+        preg_match_all(
+            '/\$this\s*->\s*block\s*\(\s*[\'"]([a-z0-9_-]+)[\'"]\s*\)/i',
+            file_get_contents($source->getPath()),
+            $matches
+        );
+
+        return new Source(
+            implode("\n", $matches[1] ?? []),
+            $source->getName(),
+            $source->getPath()
+        );
+    }
+
+    /**
+     * Check if we have the source code of a template, given its name.
+     *
+     * If we're currently in a theme context and a theme specific variant of
+     * the template exists, its availability will be checked as well.
+     *
+     * @param string $name The name of the template to check if we can load
+     *
+     * @return bool If the template source code is handled by this loader or not
+     */
+    public function exists($name): bool
+    {
+        if (parent::exists($name)) {
+            return true;
+        }
+
+        if (null !== ($themeTemplate = $this->getThemeTemplateName($name))) {
+            return parent::exists($themeTemplate);
+        }
+
+        return false;
+    }
+
+    /**
+     * Returns true if the template or any variant of it in the hierarchy is
+     * still fresh.
+     *
+     * If we're currently in a theme context and a theme specific variant of
+     * the template exists, its state will be checked as well.
+     *
+     * @param string $name The template name
+     * @param int    $time Timestamp of the last modification time of the
+     *                     cached template
+     *
+     * @throws LoaderError When $name is not found
+     *
+     * @return bool true if the template is fresh, false otherwise
+     */
+    public function isFresh($name, $time): bool
+    {
+        if ((null !== ($themeTemplate = $this->getThemeTemplateName($name))) && !parent::isFresh($themeTemplate, $time)) {
+            return false;
+        }
+
+        $chain = $this->getInheritanceChains()[$this->getIdentifier($name)] ?? [];
+
+        foreach (array_keys($chain) as $path) {
+            try {
+                if (filemtime($path) > $time) {
+                    return false;
+                }
+            } catch (\Exception $e) {
+                return false;
+            }
```

Comment:
Does `filemtime()` throw exceptions?

Reply:
Handled in 80dcb4fbd13c020ea498146743ef0c29336c2aee.